crates/xtask-bump-check/src/xtask.rs

```diff
@@ -0,0 +1,393 @@
+            if upstream_branches.len() > 1 {
+                let name = upstream_ref.name().expect("name is valid UTF-8");
+                let _ = config.shell().warn(format!(
+                    "multiple `{UPSTREAM_BRANCH}` found, picking {name}"
+                ));
+            }
+            upstream_ref.peel_to_commit()?
+        }
+    };
+    Ok(base_commit)
+}
+
+/// Returns `HEAD` of the Git repository if `head-rev` is missing.
+fn get_head_commit<'a>(
+    args: &clap::ArgMatches,
+    repo: &'a git2::Repository,
+) -> CargoResult<git2::Commit<'a>> {
+    let head_commit = match args.get_one::<String>("head-rev") {
+        Some(sha) => {
+            let head_obj = repo.revparse_single(sha)?;
+            head_obj.peel_to_commit()?
+        }
+        None => {
+            let head_ref = repo.head()?;
+            head_ref.peel_to_commit()?
+        }
+    };
+    Ok(head_commit)
+}
+
+/// Gets the referenced commit to compare if version bump needed.
+///
+/// * When merging into nightly, check the version with beta branch
+/// * When merging into beta, check the version with stable branch
+/// * When merging into stable, check against crates.io registry directly
+fn get_referenced_commit<'a>(
+    repo: &'a git2::Repository,
+    base: &git2::Commit<'a>,
+) -> CargoResult<Option<git2::Commit<'a>>> {
+    let [beta, stable] = beta_and_stable_branch(&repo)?;
+    let rev_id = base.id();
+    let stable_commit = stable.get().peel_to_commit()?;
+    let beta_commit = beta.get().peel_to_commit()?;
+
+    let referenced_commit = if rev_id == stable_commit.id() {
+        None
+    } else if rev_id == beta_commit.id() {
+        Some(stable_commit)
+    } else {
+        Some(beta_commit)
+    };
+
+    Ok(referenced_commit)
+}
+
+/// Lists all changed workspace members between two commits.
+fn changed<'r, 'ws>(
+    ws: &'ws Workspace<'_>,
+    repo: &'r git2::Repository,
+    base_commit: &git2::Commit<'r>,
+    head: &git2::Commit<'r>,
+) -> CargoResult<HashSet<&'ws Package>> {
+    let ws_members: HashMap<_, &Package> = HashMap::from_iter(
+        ws.members()
+            .filter(|pkg| pkg.publish() != &Some(vec![])) // filter out `publish = false`
+            .map(|pkg| {
+                let relative_pkg_root = pkg.root().strip_prefix(ws.root()).unwrap();
+                (relative_pkg_root, pkg)
+            }),
+    );
+    let base_tree = base_commit.as_object().peel_to_tree()?;
+    let head_tree = head.as_object().peel_to_tree()?;
+    let diff = repo
+        .diff_tree_to_tree(Some(&base_tree), Some(&head_tree), Default::default())
+        .unwrap();
+
+    let mut changed_members = HashSet::new();
+
+    for delta in diff.deltas() {
+        let mut insert_changed = |path: &Path| {
+            for (pkg_root, pkg) in ws_members.iter() {
+                if path.starts_with(pkg_root) {
+                    changed_members.insert(*pkg);
+                    break;
+                }
+            }
```

Comment:
Won't this flag everything as a change in `cargo`?  I would assume we should look for the most specific pkg_root

Reply:
Fixed by filtering out root package.

Reply:
Cheater :) 

The failure mode if we do any further nesting is just extra packages being flag for being bumped which is better than fewer packages being flagged.